File: Core/Workspace/CSharp/Adapters/Allors.Workspace.Adapters.Remote/Session/ChangeSet.cs
// <copyright file="RemoteSessionChangeSet.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>
// <summary>
//   Defines the AllorsChangeSetMemory type.
// </summary>

namespace Allors.Workspace.Adapters.Remote
{
    using System.Collections.Generic;
    using System.Linq;
    using Meta;

    internal sealed class ChangeSet : IChangeSet
    {
        public ChangeSet(Session session, ISet<IStrategy> created, ISet<IStrategy> instantiated)
        {
            this.Session = session;
            this.Created = created;
            this.Instantiated = instantiated;
            this.AssociationsByRoleType = new Dictionary<IRoleType, ISet<IStrategy>>();
            this.RolesByAssociationType = new Dictionary<IAssociationType, ISet<IStrategy>>();
        }

        ISession IChangeSet.Session => this.Session;
        private Session Session { get; }

        public ISet<IStrategy> Created { get; }

        public ISet<IStrategy> Instantiated { get; }

        public IDictionary<IRoleType, ISet<IStrategy>> AssociationsByRoleType { get; }

        public IDictionary<IAssociationType, ISet<IStrategy>> RolesByAssociationType { get; }

        internal void AddSessionStateChanges(IDictionary<IPropertyType, IDictionary<long, object>> sessionStateChangeSet)
        {
            foreach (var kvp in sessionStateChangeSet)
            {
                var ids = kvp.Value.Keys;
                var strategies = new HashSet<IStrategy>(ids.Select(v => this.Session.GetStrategy(v)));

                switch (kvp.Key)
                {
                    case IAssociationType associationType:
                        this.RolesByAssociationType.Add(associationType, strategies);
                        break;
                    case IRoleType roleType:
                        this.AssociationsByRoleType.Add(roleType, strategies);
                        break;
                }
            }
        }

        internal void AddAssociation(IRelationType relationType, Strategy association)
        {
            var roleType = relationType.RoleType;

            if (!this.AssociationsByRoleType.TryGetValue(roleType, out var associations))
            {
                associations = new HashSet<IStrategy>();
                this.AssociationsByRoleType.Add(roleType, associations);
            }

            _ = associations.Add(association);
        }

        internal void AddRole(IRelationType relationType, Strategy role)
        {
            var associationType = relationType.AssociationType;

            if (!this.RolesByAssociationType.TryGetValue(associationType, out var roles))
            {
                roles = new HashSet<IStrategy>();
                this.RolesByAssociationType.Add(associationType, roles);
            }

            _ = roles.Add(role);
        }

        internal void DiffCookedWithCooked(Strategy association, IRelationType relationType, object current, object previous)
        {
            var roleType = relationType.RoleType;

            if (roleType.ObjectType.IsUnit)
            {
                if (Equals(current, previous))
                {
                    return;
                }

                this.AddAssociation(relationType, association);
            }
            else
            {
                if (roleType.IsOne)
                {
                    if (Equals(current, previous))
                    {
                        return;
                    }

                    if (current != null)
                    {
                        this.AddRole(relationType, (Strategy)current);
                    }

                    if (previous != null)
                    {
                        this.AddRole(relationType, (Strategy)previous);
                    }

                    this.AddAssociation(relationType, association);
                }
                else
                {
                    var currentRole = (Strategy[])current;
                    var previousRole = (Strategy[])previous;

                    if (currentRole?.Length > 0 && previousRole?.Length > 0)
                    {
                        var addedRoles = currentRole.Except(previousRole);
                        var removedRoles = previousRole.Except(currentRole);

                        var hasChange = false;
                        foreach (var role in addedRoles.Concat(removedRoles))
                        {
                            this.AddRole(relationType, role);
                            hasChange = true;
                        }

                        if (hasChange)
                        {
                            this.AddAssociation(relationType, association);
                        }
                    }
                    else if (currentRole?.Length > 0)
                    {
                        foreach (var role in currentRole)
                        {
                            this.AddRole(relationType, role);
                        }

                        this.AddAssociation(relationType, association);
                    }
                    else if (previousRole?.Length > 0)
                    {
                        foreach (var role in previousRole)
                        {
                            this.AddRole(relationType, role);
                        }

                        this.AddAssociation(relationType, association);
                    }
                }
            }
        }

        internal void DiffCookedWithRaw(Strategy association, IRelationType relationType, object current, object previous)
        {
            var roleType = relationType.RoleType;

            if (roleType.ObjectType.IsUnit)
            {
                if (!Equals(current, previous))
                {
                    this.AddAssociation(relationType, association);
                }
            }
            else
            {
                if (roleType.IsOne)
                {
                    if (current == null && previous == null)
                    {
                        return;
                    }

                    var currentRole = (Strategy)current;

                    if (current != null && previous != null)
                    {
                        var previousRole = this.Session.GetStrategy((long)previous);
                        if (Equals(current, previousRole))
                        {
                            return;
                        }

                        this.AddRole(relationType, previousRole);
                        this.AddRole(relationType, currentRole);
                        this.AddAssociation(relationType, association);
                    }
                    else if (current != null)
                    {
                        this.AddRole(relationType, currentRole);
                        this.AddAssociation(relationType, association);
                    }
                    else
                    {
                        var previousRole = this.Session.GetStrategy((long)previous);
                        this.AddRole(relationType, previousRole);
                        this.AddAssociation(relationType, association);
                    }
                }
                else
                {
                    if (current == null && previous == null)
                    {
                        return;
                    }

                    var currentRole = (Strategy[])current;

                    if (previous == null)
                    {
                        foreach (var v in currentRole)
                        {
                            this.AddRole(relationType, v);
                        }

                        this.AddAssociation(relationType, association);
                    }
                    else
                    {
                        var previousRole = ((long[])previous).Select(v => this.Session.GetStrategy(v)).ToArray();

                        if (currentRole == null)
                        {
                            foreach (var v in previousRole)
                            {
                                this.AddRole(relationType, v);
                            }

                            this.AddAssociation(relationType, association);
                        }
                        else
                        {
                            var addedRoles = currentRole.Except(previousRole);
                            var removedRoles = previousRole.Except(currentRole);

                            var hasChange = false;
                            foreach (var role in addedRoles.Concat(removedRoles))
                            {
                                this.AddRole(relationType, role);
                                hasChange = true;
                            }

                            if (hasChange)
                            {
                                this.AddAssociation(relationType, association);
                            }
                        }
                    }
                }
            }
        }

        internal void DiffRawWithRaw(Strategy association, IRelationType relationType, object current, object previous)
        {
            var roleType = relationType.RoleType;

            if (roleType.ObjectType.IsUnit)
            {
                if (!Equals(current, previous))
                {
                    this.AddAssociation(relationType, association);
                }
            }
            else
            {
                if (roleType.IsOne)
                {
                    if (!Equals(current, previous))
                    {
                        if (previous != null)
                        {
                            this.AddRole(relationType, this.Session.GetStrategy((long)previous));
                        }

                        if (current != null)
                        {
                            this.AddRole(relationType, this.Session.GetStrategy((long)current));
                        }

                        this.AddAssociation(relationType, association);
                    }
                }
                else
                {
                    if (current == null && previous == null)
                    {
                        return;
                    }

                    var currentRole = (long[])current;

                    if (previous == null)
                    {
                        foreach (var v in currentRole)
                        {
                            this.AddRole(relationType, this.Session.GetStrategy(v));
                        }

                        this.AddAssociation(relationType, association);
                    }
                    else
                    {
                        var previousRole = (long[])previous;

                        if (currentRole == null)
                        {
                            foreach (var v in previousRole)
                            {
                                this.AddRole(relationType, this.Session.GetStrategy(v));
                            }

                            this.AddAssociation(relationType, association);
                        }
                        else
                        {
                            var addedRoles = currentRole.Except(previousRole);
                            var removedRoles = previousRole.Except(currentRole);

                            var hasChange = false;
                            foreach (var v in addedRoles.Concat(removedRoles))
                            {
                                this.AddRole(relationType, this.Session.GetStrategy(v));
                                hasChange = true;
                            }

                            if (hasChange)
                            {
                                this.AddAssociation(relationType, association);
                            }
                        }
                    }
                }
            }
        }

        public void DiffRawWithCooked(Strategy association, IRelationType relationType, object current, object previous)
        {
            var roleType = relationType.RoleType;

            if (roleType.ObjectType.IsUnit)
            {
                if (!Equals(current, previous))
                {
                    this.AddAssociation(relationType, association);
                }
            }
            else
            {
                if (roleType.IsOne)
                {
                    var previousRole = (Strategy)previous;
                    if (!Equals(current, previousRole?.Id))
                    {
                        if (previous != null)
                        {
                            this.AddRole(relationType, previousRole);
                        }

                        if (current != null)
                        {
                            this.AddRole(relationType, this.Session.GetStrategy((long)current));
                        }

                        this.AddAssociation(relationType, association);
                    }
                }
                else
                {
                    if (current == null && previous == null)
                    {
                        return;
                    }

                    var currentRole = (long[])current;

                    if (previous == null)
                    {
                        foreach (var v in currentRole)
                        {
                            this.AddRole(relationType, this.Session.GetStrategy(v));
                        }

                        this.AddAssociation(relationType, association);
                    }
                    else
                    {
                        var previousRole = (Strategy[])previous;

                        if (currentRole == null)
                        {
                            foreach (var v in previousRole)
                            {
                                this.AddRole(relationType, v);
                            }

                            this.AddAssociation(relationType, association);
                        }
                        else
                        {
                            var previousRoleIds = previousRole.Select(v => v.Id).ToArray();
                            var addedRoles = currentRole.Except(previousRoleIds);
                            var removedRoles = previousRoleIds.Except(currentRole);

                            var hasChange = false;
                            foreach (var v in addedRoles.Concat(removedRoles))
                            {
                                this.AddRole(relationType, this.Session.GetStrategy(v));
                                hasChange = true;
                            }

                            if (hasChange)
                            {
                                this.AddAssociation(relationType, association);
                            }
                        }
                    }
                }
            }
        }
    }
}

File: typescript/modules/libs/system/workspace/adapters/src/lib/session/change-set.ts
import { IChangeSet, IObject } from '@allors/system/workspace/domain';
import {
  AssociationType,
  PropertyType,
  RelationType,
  RoleType,
} from '@allors/system/workspace/meta';

import { MapMap } from '../collections/map-map';
import { frozenEmptySet } from '../collections/frozen-empty-set';
import { IRange, Ranges } from '../collections/ranges/ranges';

import { Session } from './session';

export class ChangeSet implements IChangeSet {
  associationsByRoleType: Map<RoleType, Set<IObject>>;
  rolesByAssociationType: Map<AssociationType, Set<IObject>>;

  private ranges: Ranges<IObject>;

  public constructor(public session: Session, public created: Set<IObject>) {
    this.associationsByRoleType = new Map();
    this.rolesByAssociationType = new Map();
    this.created ??= frozenEmptySet as Set<IObject>;

    this.ranges = this.session.ranges;
  }

  public addSessionStateChanges(
    sessionStateChangeSet: MapMap<PropertyType, IObject, unknown>
  ) {
    for (const [propertyType, map] of sessionStateChangeSet.mapMap) {
      const strategies = new Set<IObject>();

      for (const [strategy] of map) {
        strategies.add(strategy);
      }

      if (propertyType.isAssociationType) {
        this.rolesByAssociationType.set(
          propertyType as AssociationType,
          strategies
        );
      } else if (propertyType.isRoleType) {
        this.associationsByRoleType.set(propertyType as RoleType, strategies);
      } else {
        throw new Error(`PropertyType ${propertyType.name} is not supported`);
      }
    }
  }

  public diffUnit(
    association: IObject,
    relationType: RelationType,
    current: unknown,
    previous: unknown
  ) {
    if (current !== previous) {
      this.addAssociation(relationType, association);
    }
  }

  public diffCompositeStrategyRecord(
    association: IObject,
    relationType: RelationType,
    current: IObject,
    previous: number
  ) {
    if (current?.id === previous) {
      return;
    }

    if (previous != null) {
      const previousStrategy = (this.session as Session).getObject(previous);
      if (previousStrategy) {
        this.addRole(relationType, previousStrategy);
      }
    }

    if (current != null) {
      this.addRole(relationType, current);
    }

    this.addAssociation(relationType, association);
  }

  public diffCompositeRecordRecord(
    association: IObject,
    relationType: RelationType,
    current: number,
    previous: number
  ) {
    if (current === previous) {
      return;
    }

    if (previous != null) {
      this.addRole(relationType, (this.session as Session).getObject(previous));
    }

    if (current != null) {
      this.addRole(relationType, (this.session as Session).getObject(current));
    }

    this.addAssociation(relationType, association);
  }

  public diffCompositeStrategyStrategy(
    association: IObject,
    relationType: RelationType,
    current: IObject,
    previous: IObject
  ) {
    if (current === previous) {
      return;
    }

    if (previous != null) {
      this.addRole(relationType, previous);
    }

    if (current != null) {
      this.addRole(relationType, current);
    }

    this.addAssociation(relationType, association);
  }

  public diffCompositesStrategyRecord(
    association: IObject,
    relationType: RelationType,
    current: IRange<IObject>,
    previousRange: IRange<number>
  ) {
    const previous: IRange<IObject> = previousRange?.map((v) =>
      (this.session as Session).getObject(v)
    );
    this.diffCompositesStrategyStrategy(
      association,
      relationType,
      current,
      previous
    );
  }

  public diffCompositesRecordRecord(
    association: IObject,
    relationType: RelationType,
    currentRange: IRange<number>,
    previousRange: IRange<number>
  ) {
    const current: IRange<IObject> = currentRange?.map((v) =>
      (this.session as Session).getObject(v)
    );
    const previous: IRange<IObject> = previousRange?.map((v) =>
      (this.session as Session).getObject(v)
    );
    this.diffCompositesStrategyStrategy(
      association,
      relationType,
      current,
      previous
    );
  }

  public diffCompositesStrategyStrategy(
    association: IObject,
    relationType: RelationType,
    current: IRange<IObject>,
    previous: IRange<IObject>
  ) {
    let hasChange = false;

    for (const v of this.ranges.enumerate(previous)) {
      if (!this.ranges.has(current, v)) {
        this.addRole(relationType, v);
        hasChange = true;
      }
    }

    for (const v of this.ranges.enumerate(current)) {
      if (!this.ranges.has(previous, v)) {
        this.addRole(relationType, v);
        hasChange = true;
      }
    }

    if (hasChange) {
      this.addAssociation(relationType, association);
    }
  }

  private addAssociation(relationType: RelationType, association: IObject) {
    if (association == null) {
      // TODO: Investigate
      return;
    }

    const roleType = relationType.roleType;

    let associations = this.associationsByRoleType.get(roleType);
    if (!associations) {
      associations = new Set();
      this.associationsByRoleType.set(roleType, associations);
    }

    associations.add(association);
  }

  private addRole(relationType: RelationType, role: IObject) {
    if (role == null) {
      // TODO: Investigate
      return;
    }

    const associationType = relationType.associationType;

    let roles = this.rolesByAssociationType.get(associationType);
    if (!roles) {
      roles = new Set();
      this.rolesByAssociationType.set(associationType, roles);
    }

    roles.add(role);
  }
}
